packages/edit-widgets/src/components/customizer-edit-widgets-initializer/sync-customizer.js

```diff
@@ -0,0 +1,120 @@
+/**
+ * External dependencies
+ */
+import { throttle } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { parse, serialize } from '@wordpress/blocks';
+
+/*
+Widget area edits made in the Customizer are synced to Customizer
+changesets as an object, encoded as a JSON string, where the keys
+are widget area IDs and the values are serialized block content.
+This file takes care of that syncing using the 2-way data binding
+supported by `WP_Customize_Control`s. The process is as follows:
+
+- On load, the client checks if the current changeset has
+widget areas that it can parse and use to hydrate the store.
+It will load all widget areas for the current theme, but if
+the changeset has content for a given area, it will replace
+its actual published content with the changeset's.
+
+- On edit, the client updates the 2-way bound input with a new object that maps
+widget area IDs and the values are serialized block content, encoded
+as a JSON string.
+
+- On publish, a PHP action will parse the JSON string in the
+changeset and update all the widget areas in it, to store the
+new content.
+*/
+
+// Get widget areas from the store in an `id => blocks` mapping.
+const getWidgetAreasObject = () => {
+	const { getWidgetAreas, getBlocksFromWidgetArea } = window.wp.data.select(
+		'core/edit-widgets'
+	);
+
+	return getWidgetAreas().reduce( ( widgetAreasObject, { id } ) => {
+		widgetAreasObject[ id ] = getBlocksFromWidgetArea( id );
+		return widgetAreasObject;
+	}, {} );
+};
+
+// Serialize the provided blocks and render them in the widget area with the provided ID.
+const previewBlocksInWidgetArea = throttle( ( id, blocks ) => {
+	const customizePreviewIframe = document.querySelector(
+		'#customize-preview > iframe'
+	);
+	if ( ! customizePreviewIframe || ! customizePreviewIframe.contentDocument ) {
+		return;
+	}
+
+	const widgetArea = customizePreviewIframe.contentDocument.querySelector(
+		`[data-customize-partial-placement-context*='"sidebar_id":"${ id }"']`
+	);
+	if ( widgetArea ) {
+		widgetArea.innerHTML = serialize( blocks );
+	}
+}, 1000 );
+
+// Update the hidden input that has 2-way data binding with Customizer settings.
+const updateSettingInputValue = throttle( ( nextWidgetAreas ) => {
+	const settingInput = document.getElementById(
+		'_customize-input-gutenberg_widget_blocks'
+	);
+	if ( settingInput ) {
+		settingInput.value = JSON.stringify(
+			Object.keys( nextWidgetAreas ).reduce( ( value, id ) => {
+				value[ id ] = serialize( nextWidgetAreas[ id ] );
+				return value;
+			}, {} )
+		);
+		settingInput.dispatchEvent( new window.Event( 'change' ) );
+	}
+}, 1000 );
+
+// Check that all the necessary globals are present.
+if ( window.wp && window.wp.customize && window.wp.data ) {
+	// Wait for the Customizer to finish bootstrapping.
+	window.wp.customize.bind( 'ready', () =>
+		window.wp.customize.previewer.bind( 'ready', () => {
```

Comment:
I do wonder if we can attach these behaviors to an effect in the `Initializer` component for instance, I think it makes it a bit easier to reason about to think about.

Reply:
That was the original approach, but we pivoted a little bit here:

https://github.com/WordPress/gutenberg/pull/16204#discussion_r297832241